### DISCUSSION TITLE
Anyone using Decompose in production with Swift UI?

### DISCUSSION THREAD

**Discussion by chrisbanes** [arkivanov/Decompose#525] [General]
We're hitting a lot of pain right now with Decompose and the integration with Swift UI's Navigation Views (`NavigationView` or `NavigationStack`). Neither NV or NS work when nested (at all, or well), but the whole premise of Decompose is that you decompose your navigation into BLoCs, which means you're nearly always going to have nested navigation stacks.

Has anyone used Decompose + Swift UI beyond a simple sample?

It feels like Decompose and Swift UI do not particularly work well together, and I don't really see a way where they ever could.

**Comment by arkivanov** [comment]
Thanks for raising the discussion! There was similar feedback a while ago. After some research, it appears that `NavigationStack` doesn't work well when nested. E.g. [here](https://developer.apple.com/forums/thread/737671) they suggest the following:

> The intended way to use it is to have one top-level NavigationStack

Currently it looks like a limitation on the SwiftUI side. To me it looks like `NavigationStack` adds the top app bar and automatically animates its content when switching screens. It's not obvious how this could behave when nested.

The same is kinda applicable to Jetpack Compose. There is usually only one level with top app bar navigation. E.g. I can have `RootComponent` that switches its children, and each child is fullscreen and contains a top app bar. Nested children may have another level of navigation (e.g. `HomeScreen` with bottom tab bar), but there wouldn't be any top app bar in them.

That said, maybe try to only use `NavigationStack` for the fullscreen navigation? And switch children manually on nested levels?

> It feels like Decompose and Swift UI do not particularly work well together, and I don't really see a way where they ever could.

`NavigationStack` actually accepts the stack as an observable array of destinations. Which is exactly what Decompose exposes. So I think they both should work just fine. Nested navigation is indeed a nice feature of Decompose, and as I mentioned above it's subject to whether `NavigationStack` supports it or not.

**Comment by AJIEKCX** [comment]
Hi @chrisbanes, we created a wrapper on UIKit to use nested navigations stack. There is only one limitation that navigation bar exists for each screen, but all navigation gestures work. You can find the code in [this](https://gist.github.com/AJIEKCX/e1c50fb3b1c0496138bcdbb2779e3898) GitHub gist. Hope this will help you
